src/lib/pubkey/ecies/ecies.cpp

```diff
@@ -0,0 +1,385 @@
+   m_ka(key, ecies_params, false),
+   m_params(ecies_params),
+   m_iv(),
+   m_label()
+   {
+   // ISO 18033: "If v > 1 and CheckMode = 0, then we must have gcd(�, v) = 1." (v = index, � = order)
+   if(!ecies_params.check_mode())
+      {
+      Botan::BigInt cofactor = m_params.domain().get_cofactor();
+      if(cofactor > 1 && Botan::gcd(cofactor, m_params.domain().get_order()) != 1)
+         {
+         throw Invalid_Argument("ECIES: gcd of cofactor and order must be 1 if check_mode is 0");
+         }
+      }
+   }
+
+/**
+* ECIES Decryption according to ISO 18033-2
+*/
+secure_vector<byte> ECIES_Decryptor::do_decrypt(byte& valid_mask, const byte in[], size_t in_len) const
+   {
+   size_t point_size = m_params.domain().get_curve().get_p().bytes();
+   if(m_params.compression_type() != PointGFp::COMPRESSED)
+      {
+      point_size *= 2;		// uncompressed and hybrid contains x AND y
+      }
+   point_size += 1;			// format byte
+
+   std::unique_ptr<MessageAuthenticationCode> mac = m_params.create_mac();
+   BOTAN_ASSERT(mac != nullptr, "MAC is found");
+
+   if(in_len < point_size + mac->output_length())
+      {
+      throw Decoding_Error("ECIES decryption: ciphertext is too short");
+      }
+
+   // extract data
+   const std::vector<byte> other_public_key_bin(in, in + point_size);	// the received (ephemeral) public key
+   const std::vector<byte> encrypted_data(in + point_size, in + in_len - mac->output_length());
+   const std::vector<byte> mac_data(in + in_len - mac->output_length(), in + in_len);
+
+   // ISO 18033: step a
+   PointGFp other_public_key = OS2ECP(other_public_key_bin, m_params.domain().get_curve());
+
+   // ISO 18033: step b
+   if(m_params.check_mode() && !other_public_key.on_the_curve())
+      {
+      throw Decoding_Error("ECIES decryption: received public key is not on the curve");
+      }
+
+   // ISO 18033: step e (and step f because get_affine_x (called by ECDH_KA_Operation::raw_agree) 
+   // throws Illegal_Transformation if the point is zero)
+   const SymmetricKey secret_key = m_ka.derive_secret(other_public_key_bin, other_public_key);
+
+   // validate mac
+   mac->set_key(secret_key.begin() + m_params.dem_keylen(), m_params.mac_keylen());
+   mac->update(encrypted_data);
+   if(!m_label.empty())
+      {
+      mac->update(m_label);
+      }
+   const secure_vector<byte> calculated_mac = mac->final();
+   valid_mask = CT::expand_mask<byte>(same_mem(mac_data.data(), calculated_mac.data(), mac_data.size()));
+
+   // decrypt data
+   std::unique_ptr<Keyed_Filter> cipher = m_params.create_cipher(DECRYPTION);
+   BOTAN_ASSERT(cipher != nullptr, "Cipher is found");
+
+   cipher->set_key(SymmetricKey(secret_key.begin(), m_params.dem_keylen()));
+   if(m_iv.size() != 0)
+      {
+      cipher->set_iv(m_iv);
+      }
+   Pipe pipe(cipher.release());
+   pipe.process_msg(encrypted_data);
+   return pipe.read_all(0);
```

Comment:
I think only the creator of the ECIES ciphertext can create a valid MAC so I agree this does not seem to provide a usable padding oracle.

Decrypting conditionally does provide a MAC validity oracle, but this does not seem particularly harmful. Such an oracle is likely already provided when the receiver acts or fails to act on the message.

The symmetric decryption should be done in a try/catch block, if any exception is thrown during this the overall ECIES decryption fails and valid_mask is set to 0. If either the MAC check or the padding check (/GCM tag check) failed, return an empty string instead of the decrypted ciphertext.